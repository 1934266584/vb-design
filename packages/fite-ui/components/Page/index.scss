@import '../../style/theme.scss';

.mx-page {
  min-height: 100vh;
  &__head {
    width: 100%;
    &--fixed {
      position: fixed;
      top: 0;
      left: 0;
      z-index: 1000;
    }
  }
  &__foot {
    width: 100%;
    position: fixed;
    left: 0;
    bottom: 0;
    z-index: 1000;
    box-sizing: border-box;
    background: #fff;
  }
}
