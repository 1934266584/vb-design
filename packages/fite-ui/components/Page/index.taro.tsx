export default {
  name: 'page',
};
const a = '22';
console.log(a);
