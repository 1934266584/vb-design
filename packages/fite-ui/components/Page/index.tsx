import { createComponent } from '../utils';
import { nextTick, onMounted, ref, type ExtractPropTypes } from 'vue';
import { props, useOptipns } from './src/hooks';

const { create, bem } = createComponent('page');

export type PageProps = ExtractPropTypes<typeof props>;

export default create({
  props,
  setup(prop, { slots }) {
    const header = ref<HTMLDivElement>();
    const body = ref<HTMLDivElement>();
    const foot = ref<HTMLDivElement>();
    const { opts } = useOptipns();

    const getHeight = () => {
      nextTick(() => {
        opts.headerHeight = header.value?.clientHeight || 0;
        opts.footHeight = foot.value?.clientHeight || 0;
        opts.bodyHetght = document.body.clientHeight - (opts.headerHeight + opts.footHeight);
      });
    };

    onMounted(() => {
      getHeight();
    });

    return () => {
      const { fixed } = prop;
      const { bodyHetght, headerHeight } = opts;
      const style = {
        height: `${bodyHetght}px`,
        paddingTop: fixed ? `${headerHeight}px` : '0',
      };
      return (
        <div class={bem()}>
          <div
            ref={header}
            class={{
              [`${bem('head')}`]: true,
              [`${bem('head--fixed')}`]: fixed,
            }}
          >
            {slots.head?.()}
          </div>
          <div ref={body} class={bem('body')} style={style}>
            {slots.default?.()}
          </div>
          <div ref={foot} class={bem('foot')}>
            {slots.foot?.()}
          </div>
        </div>
      );
    };
  },
});
