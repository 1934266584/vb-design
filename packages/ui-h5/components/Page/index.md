# Page 分页组件

---

### 介绍

一个带有分页功能下拉刷新的的组合式功能的业务组件

### 支持框架

`Vue3`、`Taro`

### 使用

```js
import { createApp } from 'vue';

// vue
import { Page, PageList } from '@mx-design/mxui';
//taro
import { Page, PageList } from '@mx-design/mxui/mxui.taro.es';

const app = createApp();
app.use(Page);
app.use(PageList);
```

### 代码示例

#### 布局方式 1

:::demo

```html demo/base
<template>
  <div class="demo">
    <Page fixed>
      <template #head>
        <div>
          <nut-tabs v-model="state.tabValue" @click="click">
            <nut-tabpane title="上下分布"></nut-tabpane>
            <nut-tabpane title="底部悬浮按钮"></nut-tabpane>
          </nut-tabs>
        </div>
      </template>
      <template #default>
        <div>内容</div>
      </template>
    </Page>
  </div>
</template>

<script setup>
  import { Page, PageList } from '@mx-design/mxui';
  import { reactive, ref } from 'vue';

  const state = reactive({
    tabValue: 0,
  });
</script>

<style lang="scss" scoped>
  .demo {
    height: 100vh;
  }
  :deep(.nut-tabs__content) {
    display: none;
  }
</style>
```

:::

#### 布局方式 2

:::demo

```html demo/base1
<template>
  <div class="demo">
    <Page>
      <template #default>
        <div>内容</div>
      </template>
      <template #foot>
        <div class="button">
          <nut-button block type="primary">添加</nut-button>
        </div>
      </template>
    </Page>
  </div>
</template>

<script setup>
  import Page from '../index.tsx';
  import PageList from '../../PageList';
</script>

<style lang="scss" scoped>
  .demo {
    height: 100vh;
    .button {
      padding: 16px;
      background: #f5f5f5;
    }
  }
  :deep(.nut-tabs__content) {
    display: none;
  }
</style>
```

:::

#### 基本使用方式

:::demo

```html demo
<template>
  <div class="demo">
    <mx-page fixed>
      <template #head>
        <div>
          <nut-tabs v-model="state.tabValue" @click="click">
            <nut-tabpane title="分页列表"></nut-tabpane>
            <nut-tabpane title="空数据"></nut-tabpane>
          </nut-tabs>
        </div>
      </template>
      <template #default>
        <mx-page-list :fetch="fetch" ref="pageList">
          <template v-slot="{ item, index }" #default>
            <div style="height: 100px">{{ item }},{{ index }}</div>
          </template>
        </mx-page-list>
      </template>
      <template #foot>
        <div class="button">
          <nut-button block type="primary">button</nut-button>
        </div>
      </template>
    </mx-page>
  </div>
</template>

<script setup>
  import { reactive, ref } from 'vue';

  const state = reactive({
    tabValue: 0,
  });
  const pageList = ref();

  const getData = () => {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
      }, 1000);
    });
  };

  const click = ({ paneKey }) => {
    if (paneKey == 1) {
      pageList.value.clear();
    }
    if (paneKey == 0) {
      pageList.value.refresh();
    }
  };

  const fetch = async (pageSize, pageNum) => {
    const rows = await getData();
    return {
      rows,
      total: 20,
    };
  };
</script>

<style lang="scss" scoped>
  .demo {
    height: 100vh;
    .button {
      padding: 16px;
      background: #f5f5f5;
    }
  }
  :deep(.nut-tabs__content) {
    display: none;
  }
</style>
```

:::

### API

#### Page Props

| 参数  | 说明             | 类型      | 默认值 |
| ----- | ---------------- | --------- | ------ |
| fixed | 头部内容是否固定 | `Boolean` | false  |

<route>
{
  meta: {
    menu: {
      title: '业务组件',
      children: ['Page 分页组件']
    }
  }
}
</route>
