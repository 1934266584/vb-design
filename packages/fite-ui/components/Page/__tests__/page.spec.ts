import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import Page from '../index';
import { nextTick } from 'vue';

describe('Page', async () => {
  const wrapper = mount(Page);

  it('should render correctly', () => {
    nextTick(() => {
      // expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('should render fixed correctly', async () => {
    wrapper.setProps({ fixed: true });
    await wrapper.vm.$nextTick;
    expect(wrapper.find('.mx-page__head--fixed').exists()).toBeTruthy;
    // expect(wrapper.html()).toMatchSnapshot();
  });
});
