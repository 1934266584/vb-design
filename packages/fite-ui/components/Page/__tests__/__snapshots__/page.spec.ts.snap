// Vitest Snapshot v1

exports[`Page > should render correctly 1`] = `
"<div class=\\"mx-page\\">
  <div class=\\"mx-page__head\\">
    <!---->
  </div>
  <div class=\\"mx-page__body\\" style=\\"height: 0px; padding-top: 0;\\">
    <!---->
  </div>
  <div class=\\"mx-page__foot\\">
    <!---->
  </div>
</div>"
`;

exports[`Page > should render fixed correctly 1`] = `
"<div class=\\"mx-page\\">
  <div class=\\"mx-page__head mx-page__head--fixed\\">
    <!---->
  </div>
  <div class=\\"mx-page__body\\" style=\\"height: 0px; padding-top: 0px;\\">
    <!---->
  </div>
  <div class=\\"mx-page__foot\\">
    <!---->
  </div>
</div>"
`;
